### PR TITLE
Fix sudo permission checks and prevent empty file creation on error

### PR DESCRIPTION
**Pull request by isseis**
This pull request introduces improvements to user permission checks and file writing safety in the `internal/groupmembership` and `internal/safefileio` packages. The main changes are the introduction of stricter and more accurate user ID handling for permission checks (with special handling for `sudo` scenarios), enhancements to file overwrite safety (especially in error scenarios), and better test coverage for edge cases.

**User permission and UID handling improvements:**

* Added new helper functions `getPermissionCheckUID`, `parseSudoUID`, and `getProcessEUID` to accurately determine which UID should be used for permission checks, especially when running under `sudo`. This ensures that read and write operations validate permissions for the correct user, improving security and correctness.
* Updated `CanCurrentUserSafelyWriteFile` and `CanCurrentUserSafelyReadFile` to use these new helpers, replacing direct calls to `user.Current()` and manual UID parsing. This makes permission logic more robust and easier to test. [[1]](diffhunk://#diff-1f38acd458de8638d8a49f33dc7e0e5d51f7a12507f517b74718eea10fe8278bL282-R290) [[2]](diffhunk://#diff-1f38acd458de8638d8a49f33dc7e0e5d51f7a12507f517b74718eea10fe8278bL317-R324) [[3]](diffhunk://#diff-1f38acd458de8638d8a49f33dc7e0e5d51f7a12507f517b74718eea10fe8278bL348-R336)
* Added a new error variable `ErrSudoUIDOutOfRange` for better error reporting when `SUDO_UID` is invalid.

**Testing improvements:**

* Added comprehensive tests for the new UID helper functions, covering normal, sudo, invalid, and malicious `SUDO_UID` values, as well as ensuring correct behavior for both root and non-root users.

**Safe file writing and cleanup enhancements:**

* Modified the file writing logic to ensure that, if an error occurs after creating a new file, the file is removed to prevent leaving behind incomplete or invalid files. This is tracked via a `fileCreated` flag and uses the new `Remove` method on the `FileSystem` interface. [[1]](diffhunk://#diff-18f7c3b416ebe404b86cbd3a689b468b232297a7c0aa5843f30a0285ee7c77daR215-R244) [[2]](diffhunk://#diff-18f7c3b416ebe404b86cbd3a689b468b232297a7c0aa5843f30a0285ee7c77daR254-R260) [[3]](diffhunk://#diff-18f7c3b416ebe404b86cbd3a689b468b232297a7c0aa5843f30a0285ee7c77daR56-R57) [[4]](diffhunk://#diff-18f7c3b416ebe404b86cbd3a689b468b232297a7c0aa5843f30a0285ee7c77daR88-R92)
* Changed the file truncation logic: files are now truncated only after passing permission checks, and truncation is performed via a new `Truncate` method on the `File` interface for improved safety. [[1]](diffhunk://#diff-18f7c3b416ebe404b86cbd3a689b468b232297a7c0aa5843f30a0285ee7c77daR66) [[2]](diffhunk://#diff-18f7c3b416ebe404b86cbd3a689b468b232297a7c0aa5843f30a0285ee7c77daR254-R260)
* Adjusted file open flags in `safeWriteFileOverwriteWithFS` to avoid truncating files before permission validation.

These changes collectively improve the security, correctness, and reliability of permission checks and file operations in the codebase.

### CODE REVIEW

File: internal/groupmembership/manager.go [hunk 1]
```diff
@@ -449,3 +434,72 @@ func (gm *GroupMembership) clearExpiredCache() {
...
+			if parsedUID, err := strconv.Atoi(sudoUID); err == nil {
+				if parsedUID >= 0 && parsedUID <= math.MaxUint32 {
+					return parsedUID, nil
+				}
+			}
+		}
+	}
+
+	return currentUID, nil
+}
```
[the remark this answers is not in the document]
Comment (isseis) [reply internal/groupmembership/manager.go]: done

File: internal/safefileio/safe_file.go [hunk 2]
```diff
@@ -204,15 +204,33 @@ func safeWriteFileCommon(filePath string, content []byte, perm os.FileMode, fs F
...
+	// Track whether file was created by this function
+	fileCreated := false
+
 	// Use the FileSystem interface consistently for both testing and production
 	file, err := fs.SafeOpenFile(absPath, flags, perm)
 	if err != nil {
 		return err
 	}
+	// File was successfully opened/created
+	fileCreated = true
```
[the remark this answers is not in the document]
Comment (isseis) [reply internal/safefileio/safe_file.go]: done

File: internal/safefileio/safe_file.go [hunk 3]
```diff
@@ -204,15 +204,33 @@ func safeWriteFileCommon(filePath string, content []byte, perm os.FileMode, fs F
...
 
-	// Ensure the file is closed on error
+	// Ensure the file is closed, and remove it if validation or writing fails
 	defer func() {
-		if closeErr := file.Close(); closeErr != nil && err == nil {
+		closeErr := file.Close()
+
+		// If there was an error during validation or writing, remove the file
+		if err != nil && fileCreated {
+			if removeErr := os.Remove(absPath); removeErr != nil {
```
[the remark this answers is not in the document]
Comment (isseis) [reply internal/safefileio/safe_file.go]: done
